src/libraries/System.Private.CoreLib/src/System/WeakReference.T.cs

```diff
@@ -15,23 +24,64 @@ namespace System
     public sealed partial class WeakReference<T> : ISerializable
         where T : class?
     {
-        // If you fix bugs here, please fix them in WeakReference<T> at the same time.
+        // If you fix bugs here, please fix them in WeakReference at the same time.
 
         // Most methods using the handle should use GC.KeepAlive(this) to avoid potential handle recycling
         // attacks (i.e. if the WeakReference instance is finalized away underneath you when you're still
         // handling a cached value of the handle then the handle could be freed and reused).
 
-#if !CORECLR
-        // the handle field is effectively readonly until the object is finalized.
-        private IntPtr _handleAndKind;
+        private nint_t _handleAndKind;
+
+        // the lowermost 2 bits are reserved for storing additional info about the handle
+        // we can use these bits because handle is at least 32bit aligned
+        // we also reserve the sign bit
+#if CORECLR
+        private const nint_t HandleTagBits = nint_t.MinValue | 3;
+#else
+        private const nint_t HandleTagBits = 1;
+#endif
 
         // the lowermost bit is used to indicate whether the handle is tracking resurrection
-        private const nint TracksResurrectionBit = 1;
+        private const nint_t TracksResurrectionBit = 1;
+
+#if CORECLR
+        // the next bit is used to indicate whether there is an associated COM weak reference
+        private const nint_t HasComWeakReferenceBit = 2;
+        private bool HasComWeakReference() => (_handleAndKind & HasComWeakReferenceBit) != 0;
+
+        // one more bit to coordinate exclusive Set operations
+        // because of COM interop a Set may change two values, so we must ensure two Sets are not running concurrently
+        private const nint_t ExclusiveSetAccessBit = nint_t.MinValue;
```

Comment:
This unintuitive. There is nothing obvious about the minimum value of a type having a single bit set.